include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h

```diff
@@ -75,6 +76,253 @@ Optional<AnnoPathValue> resolveEntities(TokenAnnoTarget path, CircuitOp circuit,
+}
+
+/// An applier which puts the annotation on the target and drops the 'target'
+/// field from the annotaiton.  Optionally handles non-local annotations.
+template <bool allowNonLocal = false>
+static LogicalResult applyWithoutTarget(AnnoPathValue target,
+                                        DictionaryAttr anno,
+                                        ApplyState &state) {
+  return applyWithoutTargetImpl(target, anno, state, allowNonLocal);
+}
+
+struct AnnoRecord {
+  llvm::function_ref<Optional<AnnoPathValue>(DictionaryAttr, ApplyState &)>
+      resolver;
+  llvm::function_ref<LogicalResult(AnnoPathValue, DictionaryAttr, ApplyState &)>
+      applier;
+};
+
+//===----------------------------------------------------------------------===//
+// Standard Utility Resolvers
+//===----------------------------------------------------------------------===//
+
+/// Always resolve to the circuit, ignoring the annotation.
+static Optional<AnnoPathValue> noResolve(DictionaryAttr anno,
+                                         ApplyState &state) {
+  return AnnoPathValue(state.circuit);
+}
+
+/// Implementation of standard resolution.  First parses the target path, then
+/// resolves it.
+static Optional<AnnoPathValue> stdResolveImpl(StringRef rawPath,
+                                              ApplyState &state) {
+  auto pathStr = canonicalizeTarget(rawPath);
+  StringRef path{pathStr};
+
+  auto tokens = tokenizePath(path);
+  if (!tokens) {
+    mlir::emitError(state.circuit.getLoc())
+        << "Cannot tokenize annotation path " << rawPath;
+    return {};
+  }
+
+  return resolveEntities(*tokens, state.circuit, state.symTbl);
+}
+
+/// (SFC) FIRRTL SingleTargetAnnotation resolver.  Uses the 'target' field of
+/// the annotation with standard parsing to resolve the path.  This requires
+/// 'target' to exist and be normalized (per docs/FIRRTLAnnotations.md).
+static Optional<AnnoPathValue> stdResolve(DictionaryAttr anno,
+                                          ApplyState &state) {
+  auto target = anno.getNamed("target");
+  if (!target) {
+    mlir::emitError(state.circuit.getLoc())
+        << "No target field in annotation " << anno;
+    return {};
+  }
+  if (!target->getValue().isa<StringAttr>()) {
+    mlir::emitError(state.circuit.getLoc())
+        << "Target field in annotation doesn't contain string " << anno;
+    return {};
+  }
+  return stdResolveImpl(target->getValue().cast<StringAttr>().getValue(),
+                        state);
+}
+
+/// Resolves with target, if it exists.  If not, resolves to the circuit.
+static Optional<AnnoPathValue> tryResolve(DictionaryAttr anno,
+                                          ApplyState &state) {
+  auto target = anno.getNamed("target");
+  if (target)
+    return stdResolveImpl(target->getValue().cast<StringAttr>().getValue(),
+                          state);
+  return AnnoPathValue(state.circuit);
+}
+
+} // namespace
+
+const llvm::StringMap<AnnoRecord> annotationRecords{{
```

Comment:
You don't want to pull all this stuff into a header (linking and namespace issues).  Can't you just defined a
`bool isAnnoClassLowered(StringAttr class);`
as the api between LowerAnno and FIRParser?

Reply:
Cleaned up in the force-push. Awesome suggestion.